tests/buf/oom.c

```diff
@@ -1,57 +1,59 @@
 #include "clar_libgit2.h"
 #include "buffer.h"
 
-/*
- * We want to use some ridiculous size that `malloc` will fail with
- * but that does not otherwise interfere with testing.  On Linux, choose
- * a number that is large enough to fail immediately but small enough
- * that valgrind doesn't believe it to erroneously be a negative number.
- * On macOS, choose a number that is large enough to fail immediately
- * without having libc print warnings to stderr.
- */
-#if defined(GIT_ARCH_64) && defined(__linux__)
-# define TOOBIG 0x0fffffffffffffff
-#elif defined(GIT_ARCH_64)
-# define TOOBIG 0xffffffffffffff00
-#endif
-
-/**
- * If we make a ridiculously large request the first time we
- * actually allocate some space in the git_buf, the realloc()
- * will fail.  And because the git_buf_grow() wrapper always
- * sets mark_oom, the code in git_buf_try_grow() will free
- * the internal buffer and set it to git_buf__oom.
- *
- * We initialized the internal buffer to (the static variable)
- * git_buf__initbuf.  The purpose of this test is to make sure
- * that we don't try to free the static buffer.
- *
- * Skip this test entirely on 32-bit platforms; a buffer large enough
- * to guarantee malloc failures is so large that valgrind considers
- * it likely to be an error.
- */
+/* Override default allocators with ones that will fail predictably. */
+
+static git_allocator std_alloc;
+static git_allocator oom_alloc;
+
+static void *oom_malloc(size_t n, const char *file, int line)
+{
+	/* Reject any allocation of more than 100 bytes */
+	return (n > 100) ? NULL : std_alloc.gmalloc(n, file, line);
+}
+
+static void *oom_realloc(void *p, size_t n, const char *file, int line)
+{
+	/* Reject any allocation of more than 100 bytes */
+	return (n > 100) ? NULL : std_alloc.grealloc(p, n, file, line);
+}
```

Comment:
This is really cool. We might think about extending this OOM allocator to a more general helper that fails on either allocating more than `n` bytes in one allocation or if all allocations in total surpass `m` bytes. So something like `git_oomalloc_init_allocator(git_allocator *a, size_t max_allocation, size_t max_total)`. With such a utility we could easily test our code in various OOM scenarios.

One issue though with the current API is that custom allocators cannot store private data anywhere. Neither is there a field in the `git_allocator` structure that allows one to store a payload nor can users declare a "parent" structure that contains the allocator as its first field, as we do not store the global allocator as a pointer.

From these options, I'd definitely prefer having a payload pointer in the allocator structure. Making `git__allocator` a pointer creates another indirection to the function pointers that may decrease performance in allocation-heavy code